### ISSUE TITLE
Rephrase "Synchronous XMLHttpRequest" warning

### ISSUE THREAD

**Issue by suytt** [whatwg/xhr#59]
https://xhr.spec.whatwg.org/#sync-warning

> Synchronous XMLHttpRequest outside of workers is in the process of being removed from the web platform as it has detrimental effects to the end user's experience\. \(This is a long process that takes many years\.\) Developers must not pass false for the async argument when entry settings object's global object is a Window object\. User agents are strongly encouraged to warn about such usage in developer tools and may experiment with throwing an InvalidAccessError exception when it occurs\. 

This statement/deprecation is related to the way browsers do handle the (a)sync behavior.
Whether it has (or not) any "detrimental effects to the end user's experience" is subjective and should not be a valid argument.

Synchronous means it should not proceed to the next javascript statement until the request is complete, none said that browsers should freeze the UI by launching the request in the main thread...

Think about the "alert()" or "confirm()" in javascript. Those are synchronous because they block the execution of the script but it does not have "freeze" effects and does not have such "detrimental effects to the end user's experience" (subjective again).

So the specification should not be backing off the synchronous capability because browsers do implement it the wrong way.

**Comment by domenic** [comment]
> Synchronous means it should not proceed to the next javascript statement until the request is complete, none said that browsers should freeze the UI by launching the request in the main thread...

That's actually required, since e.g. changes in scroll position, clicking, etc. are detectable by JavaScript. There's no way to avoid it. That's why this is such a poorly-designed API.

> Those are synchronous because they block the execution of the script but it does not have "freeze" effects and does not have such "detrimental effects to the end user's experience" (subjective again).

They actually do have such effects. I'd welcome a test case showing otherwise, but I think you'll find that the page cannot be interacted with, scrolled, etc. while these dialogs are up.

**Comment by domenic** [comment]
Basically, since the main thread *is* the JavaScript thread, there's no way to freeze the latter without freezing the former.

**Comment by suytt** [comment]
> They actually do have such effects. I'd welcome a test case showing otherwise, but I think you'll find that the page cannot be interacted with, scrolled, etc. while these dialogs are up.

Indeed, which is the desired behavior for a synchronous call.
When you need such behavior, the alternative is to block all interactions by using DOM/CSS tricks to prevent any interaction.
What is "commonly" argued about is the entire browser entering a "not responding" state, which is not the case for alerts.

**Comment by suytt** [comment]
> Basically, since the main thread is the JavaScript thread, there's no way to freeze the latter without freezing the former.

This is implementation specific. The specification for XHR should not rely on this.
I.e.: Browsers should (1) interrupt processing of the script holding a pointer back, (2) launch the request, (3) continue script processing where it has been paused.

No where it says that synchronous calls should prevent "anything else" from happenning.

**Comment by domenic** [comment]
It is not implementation specific, actually. It is required by how all the relevant specs interact.

If you think this is wrong, I'd suggest trying to implement your proposal in a popular open source browser, and watching the many web platform tests start failing, to see the exact behaviors.

**Comment by suytt** [comment]
Dont confuse, I do not pretend to be any smarter than browser devs, just that the cited block is justified by subjective and implementation specifics.

Whether or not no vendor actually found a way to handle it properly should not alter the intent of the specification.

**Comment by domenic** [comment]
What I am trying to get across is that it is not implementation-specific, and certainly not subjective. As I've said multiple times, it's required to implement a spec-compliant browser.

**Comment by suytt** [comment]
> as it has detrimental effects to the end user's experience
This is subjective and it is the sole justification given in the warning block.

If the sync option is faulty or shall be deprecated arbitrarily, then explain how and why (or do not at all).
If the sync option is just "not desirable", this is subjective.

If the justification for deprecation of features is the browser behavior, then there should be a context-like section to explain that the scope of the specification is the browser as there are other execution environments for javascript and for XHR compliance.

**Comment by annevk** [comment]
This is a duplicate of #20.
